src/core/link_ws.go

```diff
@@ -0,0 +1,216 @@
+package core
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/Arceliar/phony"
+	"nhooyr.io/websocket"
+)
+
+type linkWS struct {
+	phony.Inbox
+	*links
+}
+
+type linkWSS struct {
+	*linkWS
+	tlsconfig  *tls.Config
+}
+
+type linkWSConn struct {
+	net.Conn
+}
+
+type linkWSListener struct {
+	ch chan *linkWSConn
+	ctx context.Context
+	httpServer *http.Server
+	listener net.Listener
+}
+
+type linkWSSListener struct {
+	ch chan *linkWSConn
+	ctx context.Context
+	httpServer *http.Server
+	listener net.Listener
+	tlslistener net.Listener
+}
+
+type wsServer struct {
+	ch chan *linkWSConn
+	ctx context.Context
+}
+
+func (l *linkWSListener) Accept() (net.Conn, error) {
+	qs := <-l.ch
+	if qs == nil {
+		return nil, context.Canceled
+	}
+	return qs, nil
+}
+
+func (l *linkWSSListener) Accept() (net.Conn, error) {
+	qs := <-l.ch
+	if qs == nil {
+		return nil, context.Canceled
+	}
+	return qs, nil
+}
+
+func (l *linkWSListener) Addr() net.Addr {
+	return l.listener.Addr()
+}
+
+func (l *linkWSSListener) Addr() net.Addr {
+	return l.listener.Addr()
+}
+
+func (l *linkWSListener) Close() error {
+	var err error = nil
+
+	err = l.httpServer.Shutdown(l.ctx)
+
+	if err != nil {
+		return err
+	}
+
+	err = l.listener.Close()
+
+	return err
+}
+
+func (l *linkWSSListener) Close() error {
+	var err error = nil
+
+	err = l.httpServer.Shutdown(l.ctx)
+
+	if err != nil {
+		return err
+	}
+
+	err = l.tlslistener.Close()
+
+	if err != nil {
+		return err
+	}
+
+	err = l.listener.Close()
+
+	return err
+}
+
+func (s *wsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	c, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		Subprotocols: []string{"ygg-ws"},
+	})
+
+	if err != nil {
+		return
```

Comment:
Do we want to return an error code in this situation? i.e.
```
  if err != nil {
    w.WriteHeader(http.StatusUpgradeRequired)
    return
  }
```
... or is the WebSocket library doing that automatically for us?

Reply:
Yes, it does one for us and this is mentioned in author's "production quality demo project" I used as source